### PR TITLE
Refactor image clients

### PR DESCRIPTION
**Pull request by djoyahoy**
- The only way to get images is now through the registry client
- Removed ggcr image and image factory
- Add image helpers for common image operations
- Add testhelpers for the registry client and keychain factory

Signed-off-by: Ashwin Venkatesh <avenkatesh@pivotal.io>

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@djoyahoy Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=pivotal/kpack&pullRequestId=247)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=pivotal/kpack&pullRequestId=247) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@djoyahoy Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=pivotal/kpack&pullRequestId=247)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=pivotal/kpack&pullRequestId=247) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@djoyahoy Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=pivotal/kpack&pullRequestId=247)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=pivotal/kpack&pullRequestId=247) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@djoyahoy Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=pivotal/kpack&pullRequestId=247)!

File: pkg/cnb/cnb_metadata.go [hunk 1]
```diff
@@ -24,67 +26,63 @@ type FetchableBuilder interface {
...
 
 func (r *RemoteMetadataRetriever) GetBuilderImage(builder FetchableBuilder) (v1alpha1.BuilderRecord, error) {
-	img, err := r.RemoteImageFactory.NewRemote(builder.Image(), registry.SecretRef{
+	builderKeychain, err := r.KeychainFactory.KeychainForSecretRef(registry.SecretRef{
 		Namespace:        builder.GetObjectMeta().GetNamespace(),
 		ImagePullSecrets: builder.ImagePullSecrets(),
 	})
 	if err != nil {
-		return emptyRecord, errors.Wrap(err, "unable to fetch remote builder baseImage")
+		return v1alpha1.BuilderRecord{}, errors.Wrap(err, "unable to create builder image keychain")
```
Comment (matthewmcnew) [inline pkg/cnb/cnb_metadata.go]: You don't like the `emptyRecord`? 😞

File: pkg/reconciler/v1alpha1/custombuilder/custombuilder_test.go [hunk 2]
```diff
@@ -90,7 +91,16 @@ func testCustomBuilderReconciler(t *testing.T, when spec.G, it spec.S) {
...
 	}
 
+	secretRef := registry.SecretRef{
+		ServiceAccount: customBuilder.Spec.ServiceAccount,
+		Namespace:      customBuilder.Namespace,
+	}
+
 	when("#Reconcile", func() {
+		it.Before(func() {
+			keychainFactory.AddKeychainForSecretRef(t, secretRef, &regtesthelpers.FakeKeychain{})
```
Comment (matthewmcnew) [inline pkg/reconciler/v1alpha1/custombuilder/custombuilder_test.go]: This can go outside of an it.Before

File: pkg/registry/client.go [hunk 3]
```diff
@@ -41,6 +53,14 @@ func (t *Client) Save(keychain authn.Keychain, tag string, image v1.Image) (stri
...
 
+func getIdentifier(image v1.Image, ref name.Reference) (string, error) {
+	digest, err := image.Digest()
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to get digest for image '%s'", ref.Context().Name())
+	}
+	return fmt.Sprintf("%s@%s", ref.Context().Name(), digest), nil
+}
+
 func previousDigest(ref name.Reference, keychain authn.Keychain, passedIn v1.Image) string {
```
Comment (matthewmcnew) [inline pkg/registry/client.go]: I realize we didn't add the fmt's here but, can we take them out here 🙂 ?

File: pkg/cnb/cnb_metadata.go (hunk 1 above)
Comment (djoyahoy) [reply pkg/cnb/cnb_metadata.go]: I feel like the idiom I've seen is to create the empty struct right then and there. Creating a const totally works, but feel like more work I suppose.

**Pull request merged**